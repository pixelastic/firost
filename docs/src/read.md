---
title: read
---

<div class="lead">
  Read the content of a file on disk and returns it as a string.
</div>

`await read(filepath)`


## Examples

```js
const content = await read('./src/css/style.css');
```

## Notes

Note that it will read the whole content before returning it, so might not be
suitable for large files (in which case you might want to use streams instead).
