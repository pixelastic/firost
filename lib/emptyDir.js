import fs from 'fs-extra';
import isDirectory from './isDirectory.js';
import absolute from './absolute.js';

/**
 * Remove content of a directory
 * @param {string} userTarget Directory to empty
 * @returns {boolean} True on success, false on error
 **/
export default async function emptyDir(userTarget) {
  const target = absolute(userTarget);
  if (!(await isDirectory(target))) {
    return false;
  }

  await fs.emptyDir(target);
  return true;
}
