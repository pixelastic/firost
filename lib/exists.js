import fs from 'fs-extra';
import { absolute } from './absolute.js';

/**
 * Check if a path exists
 * @param {string} userWhere Path to check
 * @param {object} userOptions Options to pass
 * @param {boolean} userOptions.ignoreEmptyFiles If set to true, considers empty
 * files as not existing
 * @returns {boolean} true if exists, false otherwise
 **/
export async function exists(userWhere, userOptions = {}) {
  const { ignoreEmptyFiles } = userOptions;
  const where = absolute(userWhere);
  const pathExists = await fs.pathExists(where);

  // If not found or if we don't need to check for filesize, we can return early
  if (!pathExists || !ignoreEmptyFiles) {
    return exists;
  }

  // Check that file is not empty
  const stats = await fs.stat(where);
  return stats.size > 0;
}
