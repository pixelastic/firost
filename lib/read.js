import { _ } from 'golgoth';
import fs from 'fs-extra';
import absolute from './absolute.js';

/**
 * Read any file on disk
 * @param {string} userFilepath Filepath of the file to read
 * @returns {string} Content of the file read
 **/
export default async function read(userFilepath) {
  const filepath = absolute(userFilepath);
  const content = await fs.readFile(filepath);
  return _.trim(content.toString('utf8'));
}
