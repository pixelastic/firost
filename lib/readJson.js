import read from './read.js';
import error from './error.js';

/**
 * Read a JSON file on disk and return its parsed content.
 * @param {string} source Path to the JSON file
 * @returns {Promise.<object>} The parsed content of the JSON file
 **/
export default async function readJson(source) {
  const content = await read(source);
  try {
    return JSON.parse(content);
  } catch (_err) {
    throw error('ERROR_NOT_JSON', `${source} is not a valid JSON file`);
  }
}
