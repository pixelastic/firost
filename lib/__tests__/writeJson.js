import current from '../writeJson.js';
import readJson from '../readJson.js';
import read from '../read.js';
import emptyDir from '../emptyDir.js';
import exists from '../exists.js';

describe('writeJson', () => {
  const tmpDir = './tmp/writeJson';
  beforeEach(async () => {
    await emptyDir(tmpDir);
  });
  it('should create a file with the given content', async () => {
    await current({ foo: 'bar' }, `${tmpDir}/foo.json`);

    const actual = await readJson(`${tmpDir}/foo.json`);
    expect(actual).toHaveProperty('foo', 'bar');
  });
  it('should overwrite an existing file', async () => {
    await current({ foo: 'bar' }, `${tmpDir}/foo.json`);
    await current({ foo: 'baz' }, `${tmpDir}/foo.json`);

    const actual = await readJson(`${tmpDir}/foo.json`);
    expect(actual).toHaveProperty('foo', 'baz');
  });
  it('should create nested directories', async () => {
    await current({ foo: 'bar' }, `${tmpDir}/one/two/foo.json`);

    const actual = await readJson(`${tmpDir}/one/two/foo.json`);
    expect(actual).toHaveProperty('foo', 'bar');
  });
  it('should do nothing if given undefined as content', async () => {
    await current(undefined, `${tmpDir}/file.json`);

    expect(await exists(`${tmpDir}/file.json`)).toBe(false);
  });
  it('should pretty print the output', async () => {
    await current({ foo: 'bar', bar: 'baz' }, `${tmpDir}/foo.json`);

    const actual = await read(`${tmpDir}/foo.json`);
    expect(actual).toMatchSnapshot();
  });
  it('should sort the object keys, to ease diffing', async () => {
    await current(
      { zoo: 'bar', coo: 'bar', moo: 'bar', foo: 'bar' },
      `${tmpDir}/foo.json`,
    );

    const actual = await read(`${tmpDir}/foo.json`);
    expect(actual).toMatchSnapshot();
  });
  it('should not sort the object keys with { sort: false }', async () => {
    await current(
      { zoo: 'bar', coo: 'bar', moo: 'bar', foo: 'bar' },
      `${tmpDir}/foo.json`,
      { sort: false },
    );

    const actual = await read(`${tmpDir}/foo.json`);
    expect(actual).toMatchSnapshot();
  });
  it('should prettify the output if prettier is available', async () => {
    const mockFormat = vi.fn().mockReturnValue('pretty');
    vi.spyOn(current, '__require').mockReturnValue({ format: mockFormat });
    await current({ foo: 'bar' }, `${tmpDir}/foo.json`);

    const actual = await read(`${tmpDir}/foo.json`);
    expect(actual).toBe('pretty');

    expect(current.__require).toHaveBeenCalledWith('prettier');
    expect(mockFormat).toHaveBeenCalledWith('{\n  "foo": "bar"\n}', {
      parser: 'json',
    });
  });
});
