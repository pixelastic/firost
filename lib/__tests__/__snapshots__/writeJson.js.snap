// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`writeJson > should not sort the object keys with { sort: false } 1`] = `
"{
  "zoo": "bar",
  "coo": "bar",
  "moo": "bar",
  "foo": "bar"
}"
`;

exports[`writeJson > should pretty print the output 1`] = `
"{
  "bar": "baz",
  "foo": "bar"
}"
`;

exports[`writeJson > should sort the object keys, to ease diffing 1`] = `
"{
  "coo": "bar",
  "foo": "bar",
  "moo": "bar",
  "zoo": "bar"
}"
`;

exports[`writeJson should not sort the object keys with { sort: false } 1`] = `
"{
  \\"zoo\\": \\"bar\\",
  \\"coo\\": \\"bar\\",
  \\"moo\\": \\"bar\\",
  \\"foo\\": \\"bar\\"
}"
`;

exports[`writeJson should pretty print the output 1`] = `
"{
  \\"bar\\": \\"baz\\",
  \\"foo\\": \\"bar\\"
}"
`;

exports[`writeJson should sort the object keys, to ease diffing 1`] = `
"{
  \\"coo\\": \\"bar\\",
  \\"foo\\": \\"bar\\",
  \\"moo\\": \\"bar\\",
  \\"zoo\\": \\"bar\\"
}"
`;
